### PR TITLE
Data Model Updates

### PR DESCRIPTION
**Pull request by danaspiegel**
Added updates to provide more comprehensive data models
Renamed Each->Item to be more in line with the rest of the documentation
Added a data model for Brand and a couple of other contained objects
Also updated license to ensure Lucid Green can indicate which components of API spec are copyright Lucid Green

### CODE REVIEW

File: schemas/Batch.yaml [hunk 1]
```diff
@@ -1,10 +1,33 @@
...
+  - coa_ids
+  - updated_on
 properties:
+  id:
+    $ref: Id.yaml
+  brand_id:
+    $ref: Id.yaml
+  name:
+    type: string
+  regulatory_uid:
```
Comment (elibosley) [inline schemas/Batch.yaml]: Prefer to do something similar to this: 

```  regulatorEvents:
    description: Sorted list (oldest first) of regulator events
    type: array
    items:
      $ref: Regulator.yaml
```

File: schemas/Batch.yaml [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
-  - approvalStatus
-  - approvalStatusDateTime
+  - id
+  - brand_id
+  - coa_ids
+  - updated_on
 properties:
+  id:
+    $ref: Id.yaml
+  brand_id:
```
Comment (elibosley) [inline schemas/Batch.yaml]: Noticing a lot of snake-case. I would prefer to keep json formatted with camelcase to 
1. match styles 
2. Be consistent with JSON formatted APIs 

I'd like to follow the google JSON guide for formatting these fields; https://google.github.io/styleguide/jsoncstyleguide.xml?showone=Property_Name_Format#Property_Name_Format

File: schemas/Case.yaml [hunk 3]
```diff
@@ -9,18 +9,10 @@ required:
 properties:
   id:
     $ref: Id.yaml
-  eaches:
+  items:
```
Comment (elibosley) [inline schemas/Case.yaml]: I'm not convinced that item vs eaches rename is actually the correct way to describe something. I would expect an item to be a single product, but an each might refer to a package of products sold as a single unit, thoughts on this @1e1f

File: schemas/Id.yaml [hunk 4]
```diff
@@ -1,2 +1,3 @@
 type: string
 format: uuid
+description: Unique ID for the Brand object. Format of uuid is platform dependent. Client implementations should expect that uuid values are always unique within a given implementation, but should not assume any particular format for the uuid itself other than string. Note that depending on platform implementation, it may be possible to convert the string to a valid UUID, however the specifics of such a conversion must be determined from the platform implementation.
```
Comment (elibosley) [inline schemas/Id.yaml]: Great comment! I would remove the brand object specification since these are used all over the place but the other context is fantastic!

File: schemas/Product.yaml [hunk 5]
```diff
@@ -1,20 +1,154 @@
...
   - name
-  - strainId
-  - strainName
-  - productCategoryName
-  - productCategoryType
-  - quantityType
-  - publicIngredients
-  - description
-  - isUsed
+  - updated_on
```
Comment (elibosley) [inline schemas/Product.yaml]: I don't think updated_on should be required.

File: schemas/Product.yaml [hunk 6]
```diff
@@ -1,20 +1,154 @@
...
+  - product_type
+  - usage
+  - variety
+  - package_size
```
Comment (elibosley) [inline schemas/Product.yaml]: Same with package_size / usage, I think type and variety are necessary but these seem like specific values that a product may not necessarily contain (based on the input from the brand / label vendor)

File: schemas/Product.yaml [hunk 7]
```diff
@@ -1,20 +1,154 @@
...
+  upc:
+    type: string
+  barcode:
+    type: string
+  product_line:
+    type: string
+  product_type:
+    type: string
+    enum:
+     - "blunt"
```
Comment (elibosley) [inline schemas/Product.yaml:33]: I know this may complicate things, but I think the camelcase thing should apply here too. I am pretty sure that we should keep enums in line with the spec and using camelcase.

File: schemas/Product.yaml [hunk 8]
```diff
@@ -1,20 +1,154 @@
...
+     - "sugar_free"
+  images:
+    type: object
+    properties:
+      thumbnail:
+        $ref: Image.yaml
+      primary:
+        $ref: Image.yaml
+
+  # Below are the previouly specified fields from Metrc. They are left here for reference.
```
Comment (elibosley) [inline schemas/Product.yaml]: I noticed this pattern all over. I would prefer that this specification is not reserved for lucid or something like that. We should plan on having a final specification that could potentially work for all vendors.

File: license.txt [hunk 9]
```diff
@@ -1,4 +1,4 @@
-Copyright 2022 Universal Cannabis Labeling API
+Copyright 2022 Universal Cannabis Labeling API. Some portions copyright 2021 Lucid Green, Inc.
```
Comment (elibosley) [inline license.txt:1]: This relates to a comment I added below, but since you'll probably read this first I was wondering what exactly a Lucid Green copyright means here - does this mean you own the rights to the data models, or just that you've contributed them? Does this prevent another vendor from adopting the standard and returning your data structures?

**Review by elibosley — CHANGES_REQUESTED** [review]
Dana! Thank you so much for your hard work on this - I've left some comments about the changes that I think will make this a bit more consistent, let me know your thoughts. 

@1e1f If you don't mind looking over these comments as well I would love additional feedback.

File: schemas/Batch.yaml (hunk 2 above)
Comment (1e1f) [reply schemas/Batch.yaml]: Agree, we should aim for consistency in whatever style guide we choose (I'm fine with google's json guide).

File: schemas/Batch.yaml [hunk 10]
```diff
@@ -1,10 +1,33 @@
 type: object
 required:
-  - skuId
-  - defaultLabTestingState
-  - approvalStatus
-  - approvalStatusDateTime
+  - id
+  - brand_id
+  - coa_ids
+  - updated_on
```
Comment (1e1f) [inline schemas/Batch.yaml]: This is a public-ish api and we should try to keep the required fields to a minimum as it impacts privacy the network participants. To that end, I would recommend against updated_on as a pattern.

File: schemas/Batch.yaml (hunk 1 above)
Comment (1e1f) [reply schemas/Batch.yaml]: Agree, seems clear that due to existing regulatory patterns (Metrc) that basically all regulatory id fields should be Array/List type.

File: schemas/Batch.yaml [hunk 11]
```diff
@@ -1,10 +1,33 @@
...
+  id:
+    $ref: Id.yaml
+  brand_id:
+    $ref: Id.yaml
+  name:
+    type: string
+  regulatory_uid:
+    type: string
+    format: Metrc ID
+  produced_on:
```
Comment (1e1f) [inline schemas/Batch.yaml]: Love to try and "triangulate" with Dana the dates that are actually needed by packaging and regulatory requirements and definitely serve them here. We've seen cultivated, packaged, "best by" etc. goal would be to hit a common set without too much redundancy or ambiguity.

File: schemas/Brand.yaml [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+required:
+  - id
+  - name
+properties:
+  id:
+    $ref: Id.yaml
+  name:
+    type: string
+    description: Name of the Brand
+  thumbnail:
```
Comment (1e1f) [inline schemas/Brand.yaml]: I find most of this agreeable, although I do think an Organization is a more stable noun than a Brand given that there are multiple non-brand participants in our ecosystem.

File: schemas/Case.yaml (hunk 3 above)
Comment (1e1f) [reply schemas/Case.yaml]: I would vote for each v item mostly because "item" is a relatively overloaded term and I don't believe we will ever steer this API beyond scope of CPG. Within scope of CPG I feel like Each aka EA is a rather stable and widely used concept.

File: schemas/Image.yaml [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  name:
+    type: string
+    description: File name of image.
+  url:
+    type: string
+    description: Absolute url of image that can either be downloaded or referenced in an HTML rendering context.
+  mime_type:
+    type: string
+    description: Mime type of image.
+  size_bytes:
```
Comment (1e1f) [inline schemas/Image.yaml]: Not sure an Image Type is within scope for the mvp of a mostly metadata api. However of course eventually it will be useful for retailers etc. to have this standardized. If Lucid needs to serve this data, on this endpoint sooner rather than later, can we make it any leaner?

File: schemas/Product.yaml (hunk 7 above)
Comment (1e1f) [reply schemas/Product.yaml:33]: +1

File: schemas/Manifest.yaml [hunk 14]
```diff
@@ -7,8 +7,9 @@ properties:
   manifestItems:
     type: array
     example:
-      - "qr.example.com"
-      - "qr.test.com"
+      - "https://qr.example.com/i/123465678"
```
Comment (1e1f) [inline schemas/Manifest.yaml:10]: Yes agree let's have a good example here, and possibly a link to the formatting / .well-known stuff? Using "/i" across all examples may be misleading, but I'm fine with it so long as there is documentation (want to make sure nobody accidentally hardcodes /i in their parser).

Comment (1e1f) [reply schemas/Manifest.yaml:10]: Want to remind us that the goal here is the .well-known endpoint provides parameter serialization using a standard format: https://swagger.io/docs/specification/serialization/

File: schemas/Item.yaml [hunk 15]
```diff
@@ -16,8 +16,4 @@ properties:
     type: array
     items:
       $ref: Regulator.yaml
-  batch:
```
Comment (1e1f) [inline schemas/Item.yaml]: Renaming Each to Item and removing the existing proposed joins is a pretty big change that probably requires a dedicated agenda item on a meeting rather than a PR review. Recommend further discussion.

File: schemas/Manifest.yaml (hunk 14 above)
Comment (1e1f) [inline schemas/Manifest.yaml:10]: Yes, agree we should have a good example here. Want to also suggest we make sure to be clear about documentation re: how QR's interact with .well-known via parameter serialization standards: https://swagger.io/docs/specification/serialization/

File: schemas/Size.yaml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+required:
+ - name
+ - value
+ - unit
+properties:
+  name:
+    type: string
+  value:
+    type: string
+  unit:
```
Comment (1e1f) [inline schemas/Size.yaml]: Appreciate having a first class object to represent these (CoA) measurements as it IS definitely required in our space / scope. I would push back on Size being a useful name for measurements that usually detail weight and percentage. I get that "serving size" (implemented above) is a colloquial and is perhaps where this abstraction came from? I would suggest that an "Amount" is a multi-unit label that is inclusive of weight and percentage, and further that actually the popular term serving size SHOULD be serving amount. Sorry if this seems pedantic, I don't feel that it is.

**Review by 1e1f — CHANGES_REQUESTED** [review]
Mostly opinions about nouns and joins, keeping required fields to a minimum. +1's about requiring style consistency according to some spec.

File: schemas/Batch.yaml (hunk 10 above)
Comment (danaspiegel) [reply schemas/Batch.yaml]: I'm ok with this, but the reason I have this here is to provide a way for API clients to know when information was last updated, and can make data processing decisions based upon when they last pulled this data.

File: schemas/Batch.yaml (hunk 2 above)
Comment (danaspiegel) [reply schemas/Batch.yaml]: camel case is ok.

File: schemas/Id.yaml (hunk 4 above)
Comment (danaspiegel) [reply schemas/Id.yaml]: yeah. good catch. this was written early and used after this comment was added.

File: schemas/Case.yaml [hunk 17]
```diff
@@ -9,18 +9,10 @@ required:
...
     $ref: IdArray.yaml
-  batchId:
-    $ref: Id.yaml
-  productId:
-    $ref: Id.yaml
   regulatorEvents:
```
Comment (danaspiegel) [inline schemas/Case.yaml:14]: @elibosley @1e1f not sure what a "regulatorEvent" is meant to represent. I think having a list of `regulatorUID`s can make sense here.

File: schemas/Batch.yaml (hunk 1 above)
Comment (danaspiegel) [reply schemas/Batch.yaml]: ok

File: schemas/Brand.yaml (hunk 12 above)
Comment (danaspiegel) [reply schemas/Brand.yaml]: I'm ok with this.

File: schemas/Batch.yaml (hunk 11 above)
Comment (danaspiegel) [reply schemas/Batch.yaml]: I think the only 2 that are really needed are a produced_on for a batch and packaged_on for an item

File: schemas/Case.yaml (hunk 3 above)
Comment (danaspiegel) [reply schemas/Case.yaml]: I'm ok with this.

File: schemas/Image.yaml (hunk 13 above)
Comment (danaspiegel) [reply schemas/Image.yaml]: @1e1f are you referring to mime_type? we can remove. this is intended to be helpful, but isn't a requirement.

File: schemas/Item.yaml (hunk 15 above)
Comment (danaspiegel) [reply schemas/Item.yaml]: See above re: renaming. We can leave as is. However the change to refer to batch and product by ID vs. including the actual record in the each is an efficiency thing. including the actual batch and product information in an Each record leads to a huge amount of duplication (they would be the majority of the each record if included).

Comment (danaspiegel) [reply schemas/Item.yaml]: We've found its better to refer to the batch and product records, which can be retrieved by ID as needed by the client.

File: schemas/Product.yaml (hunk 5 above)
Comment (danaspiegel) [reply schemas/Product.yaml]: agreed.

File: schemas/Product.yaml (hunk 6 above)
Comment (danaspiegel) [reply schemas/Product.yaml]: ok

File: schemas/Product.yaml (hunk 7 above)
Comment (danaspiegel) [reply schemas/Product.yaml:33]: ok

File: schemas/Product.yaml (hunk 8 above)
Comment (danaspiegel) [reply schemas/Product.yaml]: @elibosley not sure I understand your comment. can you explain?

File: schemas/Size.yaml (hunk 16 above)
Comment (danaspiegel) [reply schemas/Size.yaml]: ok.

**Review by danaspiegel — COMMENTED** [review]
with the exception of the question about copyright, I think I've responded to all of your comments. I'll make updates to this PR accordingly.

Copyright is, I think, a separate discussion.

File: schemas/Product.yaml (hunk 8 above)
Comment (elibosley) [reply schemas/Product.yaml]: Yeah - I'd like to remove any duplicated fields from metrc that might be reproduced by the spec that you added - that way we're not over-complicating the specification.

File: schemas/Item.yaml (hunk 15 above)
Comment (elibosley) [reply schemas/Item.yaml]: I understand your concern here- any reason we can't add a headers or query param to expand those - something like expandFields="batch,whatever" ?

File: schemas/Case.yaml (hunk 17 above)
Comment (1e1f) [reply schemas/Case.yaml:14]: Yes they are synonyms for sure. "regulatorEvent" is simply a UID string along with a "kind". The UID is of course the main requirement, the "kind" is a type hint for which transform (such as a transfer) has yielded the child tag. See Regulator.yaml

File: schemas/Batch.yaml [hunk 18]
```diff
@@ -1,67 +1,27 @@
...
+    type: string
+    format: date
+  coaIds:
+    type: array
+    items:
+      type: Id.yaml
+  updatedOn:
     type: string
     format: date-time
-    description: Uses same name as Metrc ApprovalStatusDateTime
```
Comment (elibosley) [inline schemas/Batch.yaml:19]: @danaspiegel Why remove all of these content fields?

File: schemas/Batch.yaml [hunk 19]
```diff
@@ -1,67 +1,27 @@
...
-    enum: ["Approved"]
-    description: Uses same name as Metrc ApprovalStatus
-  approvalStatusDateTime:
+  id:
+    $ref: Id.yaml
+  organizationId:
+    $ref: Id.yaml
+  name:
+    type: string
+  regulatoryUids:
```
Comment (elibosley) [inline schemas/Batch.yaml]: Rename to regulatorEvents

File: schemas/Batch.yaml (hunk 18 above)
Comment (elibosley) [reply schemas/Batch.yaml:19]: We should probably have some information here or a COA endpoint to join this with to get information about COAs.

File: schemas/Product.yaml [hunk 20]
```diff
@@ -1,108 +1,146 @@
-required:
+# copyright 2021 Lucid Green, Inc.
```
Comment (elibosley) [inline schemas/Product.yaml]: Please remove this or match the main copyright statement

File: schemas/Product.yaml [hunk 21]
```diff
@@ -1,108 +1,146 @@
...
+     - "honey"
+     - "infusedBeverage"
+     - "inhaler"
+     - "joint"
+     - "kief"
+     - "lube"
+     - "preGround"
+     - "preRolls"
+     - "rollerBallGelPen"
+     - "rso_oil"
```
Comment (elibosley) [inline schemas/Product.yaml]: camelcase this

File: schemas/Batch.yaml (hunk 18 above)
Comment (danaspiegel) [reply schemas/Batch.yaml:19]: @elibosley the reason to remove these is that they don't belong at the Batch level. They are tested data fields, and there may be multiple sets of results per batch. 

for ex. a 0.5g and 1g vape pen from the same Batch would each have a COA, with slightly different results (due to testing variability and possibly differing lab techniques). A Batch itself may have its own set of COA results, but also might not. And each packaged product may get another set of COA results through distributor testing in certain instances.

Basically, Batches represent production processes with intermediate outputs, and the final products themselves have the actual tested compound values.
